lib/SILGen/SILGenDestructor.cpp

```diff
@@ -219,6 +220,135 @@ void SILGenFunction::destroyClassMember(SILLocation cleanupLoc,
   }
 }
 
+llvm::SmallSetVector<VarDecl*, 4> findRecursiveLinks(ClassDecl *cd) {
+  auto SelfTy = cd->getDeclaredInterfaceType();
+
+  // Collect all stored properties that would form a recursive structure,
+  // so we can remove the recursion and prevent the call stack from
+  // overflowing.
+  llvm::SmallSetVector<VarDecl*, 4> recursiveLinks;
+  for (VarDecl *vd : cd->getStoredProperties()) {
+    auto Ty = vd->getInterfaceType()->getOptionalObjectType();
+    if (Ty && Ty->getCanonicalType() == SelfTy->getCanonicalType()) {
+      recursiveLinks.insert(vd);
+    }
+  }
+
+  // NOTE: Right now we only optimize linear recursion, so if there is more than one link,
+  // clear out the set and don't perform any recursion optimization.
+  if (recursiveLinks.size() > 1) {
+    recursiveLinks.clear();
+  }
+
+  return recursiveLinks;
+}
+
+
+void SILGenFunction::emitRecursiveChainDestruction(ManagedValue selfValue,
+                                                   ClassDecl *cd,
+                                                   SmallSetVector<VarDecl*, 4> recursiveLinks,
+                                                   CleanupLocation cleanupLoc) {
+  auto SelfTy = F.mapTypeIntoContext(cd->getDeclaredInterfaceType());
+
+  assert(recursiveLinks.size() <= 1 && "Only linear recursion supported.");
+
+  auto SelfTyLowered = getTypeLowering(SelfTy).getLoweredType();
+  for (VarDecl* vd : recursiveLinks) {
```

Comment:
Yes, the idea was for this to signify that the intention is to eventually optimize branching cases as well. But thinking about it, it would likely be separate logic, so I think it would make sense to simplify it here and remove the loop.